pyhap/characteristic.py

```diff
@@ -98,34 +94,66 @@ def __init__(self, display_name, type_id, properties, value=None, broker=None):
             - else use `HAP_FORMAT.DEFAULT` for the format of this characteristic.
         :type value: Depends on `properties["Format"]`
         """
-        assert "Format" in properties and "Permissions" in properties
+        if 'Format' not in properties or 'Permissions' not in properties:
+            raise ValueError('Invalid properties')
         self.display_name = display_name
         self.type_id = type_id
         self.properties = properties
-        if value:
-            self.value = value
-        else:
-            if self.properties.get('ValidValues'):
-                self.value = min(self.properties['ValidValues'].values())
-            else:
-                self.value = HAP_FORMAT.DEFAULT[properties["Format"]]
         self.broker = broker
         self.setter_callback = None
+        self.value = value or self._get_default_value()
 
     def __repr__(self):
         """Return the representation of the characteristic."""
-        return "<characteristic display_name='{}' value={} properties={}>" \
+        return '<characteristic display_name={} value={} properties={}>' \
             .format(self.display_name, self.value, self.properties)
 
+    def _get_default_value(self):
+        """Helper method. Return default value for format."""
+        if self.properties.get('ValidValues'):
+            return min(self.properties['ValidValues'].values())
+        else:
+            value = HAP_FORMAT.DEFAULT[self.properties['Format']]
+            return self._validate_value(value)
+
+    def _validate_value(self, value):
```

Comment:
This method actually transforms the value to a valid one, i.e. it is not validation only. Can we rename to `to_valid_value` or something like this?

Reply:
Actually I think this would be quite a good usecase for `@property` and `@property.setter` to do `to_valid_value`, since this should be executed every time we try to set a value.

Reply:
I know I suggested an `@property.setter` above. But thinking about it some more, I am a little concerned that using a `@property.setter` might result in confusing behaviour.

As a user of the API you would write a value to `.value`, and it could either raise an `exception` or when you read `.value` you wouldn't get the value you expected.

Just wondering if instead `.value` should be stored in `._value`, and then expose a `set_value()` and `update_value()` methods which would do the validation. Also using a `@property` getter to allow reading `.value` but not writing to it.

That way, to a user it's clear that when you call `set_value()`/`update_value()` other clever things are gonna happen (including callbacks and notifications). So should be surprised if the value passed is changed, or a validation exception is raised.

Reply:
After my comment I though about this again an I would agree in many cases @thomaspurchas:
 - a property.setter might indeed be confusing
 - ._value might be a good idea. We use the value variable only as cache for to_HAP if we need to send this to HK. I don't think the user should interact with value directly. _value should be enough, we don't need to explicitly constrain it.
 - @thomaspurchas I understand were your push for two separate methods is comming from. I personally don't like the way it't currently handled. Since should_notify is set to false when update came from HomeKit client, the callback needs to call set_value, or at least notify, so  that all clients have the updated value. Additionally when you want to call set_value, I always need to pass in callback=false to prevent the callback. That is kind of redundant.